python/ray/tune/experiment/trial.py

```diff
@@ -898,6 +911,21 @@ def __getstate__(self):
         state["_state_valid"] = False
         state["_default_result_or_future"] = None
 
+        # Save persistent trial checkpoint data (with relative paths)
+        # Upon load, the path should be constructed again relative to the
+        # `logdir`, which might be updated.
+        checkpoint_states = []
+        for checkpoint in self.get_trial_checkpoints():
+            checkpoint_state = checkpoint.__dict__.copy()
+            checkpoint_dir = checkpoint.dir_or_data
+            if isinstance(checkpoint_dir, ray.ObjectRef):
+                checkpoint_dir = ray.get(checkpoint_dir)
+            checkpoint_state["__relative_dir"] = os.path.relpath(
+                checkpoint_dir, self.logdir
+            )
+            checkpoint_states.append(checkpoint_state)
+        state["__checkpoints"] = checkpoint_states
```

Comment:
This is generally good, but it does mean that we store the same information twice - once here and once in the checkpoint manager.

To resolve this, a few options

1. Only store relative logdirs in e.g. `__checkpoint_relative_dirs` or so instead of the whole state. Then only override this on restore
2. Store the full state here, but do not store the state in the checkpoint manager

```
# getstate
checkpoint_manager_state = state["checkpoint_manager"]
checkpoints = checkpoint_manager_state.pop("_top_persisted_checkpoints")
for checkpoint in checkpoints:
    # ...
state["__checkpoint_states"] = checkpoint_states

# setstate
checkpoint_manager_state = state["checkpoint_manager"]
for checkpoint in state.pop("__checkpoint_states", []):
    top_persisted_checkpoints = ...
checkpoint_manager_state["_top_persisted_checkpoints"] = top_persisted_checkpoints
```


happy with either option but would like to avoid storing the same thing in multiple locations. Slightly tending towards the second option even though it refers to internal APIs of the checkpoint manager. What do you think?

Reply:
I took a stab at 2 but felt that it added a lot of extra complexity for not much benefit. The main issue I ran into was that `_top_persisted_checkpoints` contains `_HeapCheckpointWrapper` objects, and I had to create new `_TrackedCheckpoint` objects and insert with `handle_checkpoint`. This would also keep incrementing from the loaded `_latest_checkpoint_id`.

I think it makes more sense to go with the first option here, where we can keep the checkpoint information in the manager, then just store the extra relative dir information.

Reply:
Great, thanks for the context, this makes a lot of sense to me!

Okay, let's go with option 1 then